rclpy/rclpy/executors.py

```diff
@@ -0,0 +1,290 @@
+# Copyright 2017 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from concurrent.futures import ThreadPoolExecutor as _ThreadPoolExecutor
+import multiprocessing
+
+from rclpy.constants import S_TO_NS
+from rclpy.impl.implementation_singleton import rclpy_implementation as _rclpy
+from rclpy.utilities import ok
+
+
+class Executor:
+    """
+    A base class for an executor.
+
+    An executor controls the threading model used to process callbacks. Callbacks are units of work
+    like subscription callbacks, timer callbacks, service calls, and received client responses. An
+    executor controls which threads callbacks get executed in.
+
+    A custom executor must define the method spin_once(). If the executor has any special cleanup
+    it should also define the method shutdown().
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._nodes = set()
+        # This is triggered when wait_for_ready_callbacks should rebuild the wait list
+        gc, gc_handle = _rclpy.rclpy_create_guard_condition()
+        self._guard_condition = gc
+        self._guard_condition_handle = gc_handle
+
+    def shutdown(self, timeout_sec=None):
+        """
+        Stop executing callbacks and wait for their completion.
+
+        timeout_sec - seconds to wait for callbacks to complete. Blocks forever if None. Returns
+                  immediately if <= 0
+        """
+        _rclpy.rclpy_destroy_guard_condition(self._guard_condition)
+
+    def add_node(self, node):
+        """
+        Add a node whose callbacks should be managed by this executor.
+
+        return true if the node was added
+        """
+        self._nodes.add(node)
```

Comment:
What happens if you add a node while spinning in another thread? Don't you need to interrupt any running spins after adding the node? Also don't you need to make this function thread-safe?

Reply:
I didn't consider that. Yes to both questions, it should be made thread safe and trigger a rebuild of the wait set.

Reply:
Fixed in 89ed2c4fcad9f2a94aa23562476e6d340bacde43. The commit also eliminates a case where `wait_for_ready_callbacks` could return before the timeout was reached without yielding any work. Without fixing that case `add_node` would cause the next `spin_once` to return without doing anything because the wait list guard condition was triggered. This broke the executor tests.